File: chapters/3.4/main.py
import httpx
import serial.tools.list_ports
from serial import Serial

student = input('学号：')


def bytes2string(_data):
    return ''.join([f'{ch:0>2x}' for ch in _data]).upper()


data = b'\xaa\x55' + bytes([ord(ch) - ord('0') for ch in student])
pipe = Serial(list(serial.tools.list_ports.comports()[0])[0], 1200)

serial, passwd = bytes2string(pipe.read(13)), None
print(f'序列号：{serial}')

pipe.write(data)
for i in range(260):
    pipe.read(2)                            # 读掉 AA 55
    distance = list(pipe.read(1))[0]        # 读取起始字节位置
    pipe.read(distance - 4)                 # 读掉多余的字节
    passwd = pipe.read(4)                   # 读取 4 字节的「密码」
    data = bytes2string(passwd)             # 将「密码」格式化为易读的格式
    print(f'[{i:>3d}]: {data}', end='\r')
    pipe.write(b'\xaa\x55' + passwd)        # 将「密码」发送回单片机

print('\n')
resp = httpx.get(f'http://132.232.98.70:6363/checkSecret?id={student}&s={data}&v={serial}')
print(f'已提交至服务器，有效成绩：{resp.text}')

File: chapters/3.4/main.cpp
#include <cstdio>
#include <iostream>
#include <string>

#include "serial.h"

std::string format(const std::vector<uchar> &data) {
    std::string str(2 * data.size() + 1, '\x00');
    for (int i = 0; i < data.size(); i++) {
        sprintf(&str[i * 2], "%02X", data[i]);
    }
    return str;
}

int main() {
    printf("> ");
    std::string student; std::cin >> student;
    const std::vector<uchar> prefix = { 0xAA, 0x55 };

    serial pipe("/dev/ttyUSB0", B1200);

    printf("%s\n", format(pipe.read(13)).c_str());

    std::vector<uchar> data = prefix;
    for (char ch : student) {
        data.push_back(ch - '0');
    }
    pipe.write(data);

    for (int i = 0; i < 260; i++) {
        pipe.read(2);                                    // 读掉 AA 55
        int position = pipe.read(1)[0];                  // 读取起始字节位置
        pipe.read(position - 4);                         // 读掉多余的字节
        data = pipe.read(4);                             // 读取 4 字节的「密码」

        std::string passwd = format(data);               // 将「密码」格式化为易读的格式
        printf("[%3d]: %s\r", i, passwd.c_str());
        fflush(stdout);                                  // 刷新输出缓冲区

        // 将 AA 55 插入到收到的「密码」之前
        data.insert(data.begin(), prefix.begin(), prefix.end());
        pipe.write(data);                                // 将「密码」发送回单片机
    }
    printf("\n");

    return 0;
}
